go-controller/pkg/controller/controller.go

```diff
@@ -214,3 +237,97 @@ func (c *controller[T]) ReconcileAll() {
 		c.queue.Add(key)
 	}
 }
+
+// SubController is a controller that should be used as a part of MultiController.
+// It is very similar to Controller, but it doesn't have its own initialSync method.
+// Initial sync is handled by MultiController instead.
+// SubController is not meant to be used standalone, all its methods are private and will be called by MultiController.
+type SubController interface {
+	addHandler() error
+	startWorkers(threadiness int) error
+	stop()
+}
+
+func NewSubController[T any](name string, config *Config[T]) SubController {
+	return &controller[T]{
+		name:   name,
+		config: config,
+		queue: workqueue.NewRateLimitingQueueWithConfig(
+			config.RateLimiter,
+			workqueue.RateLimitingQueueConfig{
+				Name: name,
+			},
+		),
+		stopChan: make(chan struct{}),
+		wg:       &sync.WaitGroup{},
+	}
+}
+
+// MultiController manages multiple controllers.
+// It is only needed when initial sync is required across multiple informers.
+// If your intialSync method `List`s more than one object type, it is very likely you should use MultiController.
+// InitialSync will be run after all subcontroller event handlers are added to ensure proper event handling
+// for related informers.
+type MultiController interface {
```

Comment:
I don't understand the need of `when initial sync is required across multiple informers`. I haven't seen that need before. Could you explain why would we have than need further? 

Is it just that we need initial syncs to be called before we start event handling? So it is not that we need a single sync but an order thing?

In that case, I don't think we need a new controller type, rather an utility that is able to start multiple controllers at the same time (and potentially the only way to start controllers, removing the Start() method from the Controller interface). The SubController private interface could just be part of the Controller interface that this utility could use.

Reply:
it is used currently in some controllers, but may be not obvious. The general point of non-racy initial sync, is that you start informers, and they start adding changes object keys to your queue. Then you do initial sync by getting a `List` of objects, do the initial sync like db cleanup or cache initialization, and only after that you start handling queue keys that were being added this whole time by informers. The ordering here is important, because as soon you do `List` you don't want to miss any events happening after that.
Now if you have multiple k8s object that are related, your initial sync may need to `List` objects of different types. that means that you should start informers for ALL of these objects before initial sync, and all handlers after.
Some examples of existing syncs for multiple objects:
egressservice https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go#L228
apbroute https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/apbroute/repair.go#L30
This change is specifically needed for https://github.com/ovn-org/ovn-kubernetes/pull/4045/files#diff-d50c01ea37b6de8224c3dff25787af74bf8a440abd23c60899c69a83c0b94e96R137 that is not merged yet.
I also expect that there are a couple more controllers that are just racy and don't use initial sync properly.